### PR TITLE
Allow multiple keys for decoding/key rotation

### LINKED ISSUE #67: Document using multiple codecs in the readme
**Is your feature request related to a problem? Please describe.**
> I'd like the ability to more easily rotate keys by being able to add a list of valid decryption keys.

**Describe the solution you'd like**
>Change 
`New(hashKey, blockKey []byte)` to 
`New(hashKey, blockKey []byte, decodeKeys ...[]byte)` and add a check that if a cookie cannot be decoded with the hashKey to try with the decode keys until one matches or all fail. This would allow users to have some logic like 
`New(todaysKey(), nil, tomorrowsKey(), yesterdaysKey())` and rotate keys automatically even in a distributed environment. It would also preserve backwards compatibility with `New(hashKey, nil)`.

**Describe alternatives you've considered**
> There's the obvious, just don't rotate the keys. Or coordinate a failover mechanism by chaining decodes using multiple secure cookies. This seems like a more straightforward approach however.

### PR DESCRIPTION
**Pull request by tflyons**
Fixes #67

**Summary of Changes**

1. Adds decodeKeys... to `New()` function
2. If hashKey fails during decoding, check against decodeKeys
3. Add testing for the above

### CODE REVIEW

File: securecookie.go [hunk 1]
```diff
@@ -129,17 +129,21 @@ type Codec interface {
...
 // The default encoder used for cookie serialization is encoding/gob.
 //
+// decodeKeys are optional, used to authenticate during decoding if the hashKey
+// fails. This allows older hashKeys to be verified during routine key rotation.
+//
 // Note that keys created using GenerateRandomKey() are not automatically
 // persisted. New keys will be created when the application is restarted, and
 // previously issued cookies will not be able to be decoded.
-func New(hashKey, blockKey []byte) *SecureCookie {
+func New(hashKey, blockKey []byte, decodeKeys ...[]byte) *SecureCookie {
```
Comment (elithrar) [inline securecookie.go]: This breaks the public API - variadic arguments still change the function signature.

File: securecookie.go [hunk 2]
```diff
@@ -129,17 +129,21 @@ type Codec interface {
...
 	s := &SecureCookie{
-		hashKey:   hashKey,
-		blockKey:  blockKey,
-		hashFunc:  sha256.New,
-		maxAge:    86400 * 30,
-		maxLength: 4096,
-		sz:        GobEncoder{},
+		hashKey:    hashKey,
+		decodeKeys: decodeKeys,
```
Comment (elithrar) [inline securecookie.go]: If you wanted to do this, you could just add a method that allows adding new (named) keys to a map of keys on the SecureCookie type.

Comment (tflyons) [reply securecookie.go]: I'm entirely on board for instead adding a method to add new named keys to a map, I'll put something together today

**Comment by tflyons** [comment]
> Why not https://godoc.org/github.com/gorilla/securecookie#CodecsFromPairs - which supports this today?
> 
> Do we need to document it further?

This may work for my intentions, but it could benefit from more docs. I'll try it out today and get back to you with any doc suggestions I can come up with

**Comment by tflyons** [comment]
Closing this out, the MultiEncode/MultiDecode functions are sufficient but require better docs. I'll open a separate PR to that effect

**Pull request closed without merging**